compiler/rustc_resolve/src/build_reduced_graph.rs

```diff
@@ -406,6 +406,97 @@ impl<'a, 'ra, 'tcx> BuildReducedGraphVisitor<'a, 'ra, 'tcx> {
         self.r.field_names.insert(def_id, fields);
     }
 
+    fn resolve_restriction(&mut self, restriction: &ast::Restriction) -> ty::Restriction {
+        self.try_resolve_restriction(restriction, true).unwrap_or_else(|err| {
+            self.r.report_restriction_error(err);
+            ty::Restriction::Unrestricted
+        })
+    }
+
+    fn try_resolve_restriction<'ast>(
+        &mut self,
+        restriction: &'ast ast::Restriction,
+        finalize: bool,
+    ) -> Result<ty::Restriction, RestrictionResolutionError<'ast>> {
+        let parent_scope = &self.parent_scope;
+        match restriction.kind {
+            // If the restriction is implied, it has no effect when the item is otherwise visible.
+            ast::RestrictionKind::Unrestricted | ast::RestrictionKind::Implied => {
+                Ok(ty::Restriction::Unrestricted)
+            }
+            ast::RestrictionKind::Restricted { ref path, id, shorthand: _ } => {
+                // For restrictions we are not ready to provide correct implementation of "uniform
+                // paths" right now, so on 2018 edition we only allow module-relative paths for now.
+                // On 2015 edition visibilities are resolved as crate-relative by default,
+                // so we are prepending a root segment if necessary.
+                let ident = path.segments.get(0).expect("empty path in restriction").ident;
+                let crate_root = if ident.is_path_segment_keyword() {
+                    None
+                } else if ident.span.is_rust_2015() {
+                    Some(Segment::from_ident(Ident::new(
+                        kw::PathRoot,
+                        path.span.shrink_to_lo().with_ctxt(ident.span.ctxt()),
+                    )))
+                } else {
+                    return Err(RestrictionResolutionError::Relative2018(ident.span, path));
+                };
+
+                let segments = crate_root
+                    .into_iter()
+                    .chain(path.segments.iter().map(|seg| seg.into()))
+                    .collect::<Vec<_>>();
+                let expected_found_error = |res| {
+                    Err(RestrictionResolutionError::ExpectedFound(
+                        path.span,
+                        Segment::names_to_string(&segments),
+                        res,
+                    ))
+                };
+                match self.r.resolve_path(
+                    &segments,
+                    Some(TypeNS),
```

Comment:
The `visibility` counterpart seems to be using `None` here, should we do the same?
https://github.com/rust-lang/rust/blob/12ce9a93d84cb0bbaa8b57a26e38800cbf2841bd/compiler/rustc_resolve/src/build_reduced_graph.rs#L354

Reply:
I see the comment ``// `None` indicates a module path in import``, so that's definitely correct.

Reply:
Just to make it clear, what is correct here? Having `Some(TypeNS)` or `None`?